src/genkey/clu_genkey.c

```diff
@@ -146,140 +149,273 @@ int wolfCLU_genKey_ED25519(WC_RNG* rng, char* fOutNm, int directive, int format)
+        }
+        else {
+            derSz = wc_EccPublicKeyToDer(key->internal, der, derSz, 1);
+            if (derSz < 0) {
+                ret = derSz;
+            }
+        }
+    }
+
+    if (ret == 0) {
+        ret = wolfSSL_BIO_write(out, der, derSz);
+        if (ret != derSz) {
+            ret = -1;
+        }
+    }
+
+    if (der != NULL)
+        XFREE(der, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+
+    return ret;
+}
+
+
+/* returns 0 on successfully writing out private DER key */
+static int wolfCLU_ECC_write_priv_der(WOLFSSL_BIO* out, WOLFSSL_EC_KEY* key)
+{
+    int derSz = 0, ret = 0;
+    unsigned char *der = NULL;
+
+    if (out == NULL || key == NULL)
+        return BAD_FUNC_ARG;
+
+    ret = wc_EccKeyDerSize(key->internal, 0);
+    if (ret > 0) {
+        derSz = ret;
+        der   = (byte*)XMALLOC(derSz, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+        if (der == NULL) {
+            ret = MEMORY_E;
+        }
+        else {
+            ret = wc_EccPrivateKeyToDer(key->internal, der, derSz);
+        }
+    }
+
+    if (ret > 0) {
+        printf("writing out %d bytes for private key\n", derSz);
+        ret = wolfSSL_BIO_write(out, der, derSz);
+        if (ret != derSz) {
+            ret = -1;
+        }
+        printf("ret of write = %d\n", ret);
+    }
+
+    if (der != NULL)
+        XFREE(der, HEAP_HINT, DYNAMIC_TYPE_TMP_BUFFER);
+
+    if (ret > 0)
+        ret = 0; /* successfully wrote out key */
+
+    return ret;
+}
+#endif /* HAVE_ECC */
+
 int wolfCLU_genKey_ECC(WC_RNG* rng, char* fName, int directive, int fmt,
-                       int keySz)
+                       char* name)
 {
 #ifdef HAVE_ECC
-    ecc_key key;
-    FILE*   file;
-    int     ret;
-
     int   fNameSz     = XSTRLEN(fName);
     int   fExtSz      = 6;
     char  fExtPriv[6] = ".priv\0";
     char  fExtPub[6]  = ".pub\0\0";
     char* fOutNameBuf = NULL;
-    
-    #ifdef NO_AES
-    /* use 16 bytes for AES block size */
-    size_t maxDerBufSz = 4 * keySz * 16;
-    #else
-    size_t maxDerBufSz = 4 * keySz * AES_BLOCK_SIZE;
-    #endif
-    byte*  derBuf      = NULL;
-    int    derBufSz    = -1;
+
+    WOLFSSL_BIO *bioOut = NULL;
+    WOLFSSL_BIO *bioPub = NULL;
+    WOLFSSL_BIO *bioPri = NULL;
+    WOLFSSL_EC_KEY   *key;
+    int ret = 0;
+
+    byte*  der   = NULL;
+    int    derSz = -1;
 
     if (rng == NULL || fName == NULL)
         return BAD_FUNC_ARG;
 
-    if (fmt == PEM_FORM) {
-        printf("Der to Pem for rsa key not yet implemented\n");
-        printf("FEATURE COMING SOON!\n");
-        return FEATURE_COMING_SOON;
-    }
-
-    ret = wc_ecc_init_ex(&key, HEAP_HINT, INVALID_DEVID);
-    if (ret != 0)
-        return ret;
-    ret = wc_ecc_make_key(rng, keySz, &key);
-#if defined(WOLFSSL_ASYNC_CRYPT)
-    /* @Audit: is this all correct? */
-    ret = wc_AsyncWait(ret, &key.asyncDev, WC_ASYNC_FLAG_CALL_AGAIN);
-#endif
-    if (ret != 0)
+    key = wolfSSL_EC_KEY_new();
+    if (key == NULL)
```

Comment:
Need to return `MEMORY_E` or similar instead of `return ret`, which would return 0.